### DISCUSSION TITLE
Working properly on PS4 with the Raspberry Pi 3A+

### DISCUSSION THREAD

**Discussion by SubatomicPlanets** [Berny23/LD-ToyPad-Emulator#142] [Show and tell]
Just wanted to let you know that I think I managed to get it working on the 3A+ because the description says it does not work. As you can see, it is very similar. In my case I have a PS4, but it might also work on other consoles. If you have any questions, or so, feel free to ask. Here is the updated instruction for the RPI 3a+

**Prerequisites**

- **Raspberry Pi 3A+**
  - **NOTE**: This is not really specific to the 3A+ so it may also work on the RPI 2, 3, 3B and so on, but I only tested it on the 3A+
- **USB cable Type A to Type A** this cable is a bit rare, but can be found on amazon.
- 2 GB+ Micro SD card
- Internet connection on your PC

**Guide**

1. Flash Raspberry Pi OS Lite to your SD card using [the Raspberry Pi Imager tool](https://www.raspberrypi.org/software/). Configure the settings in the raspberry pi imager to use your wifi network, enable ssh and configure the settings you like.
 
**Settings Image:**

![RPI](https://github.com/Berny23/LD-ToyPad-Emulator/assets/133254461/21999779-08eb-468f-b315-e34e514a6e7f)

2. Connect the Raspberry Pi to your PC via USB cable

3. Use SSH to run the following commands (Don't know the IP address? Try "raspberrypi.local" instead of an IP adress.)

```
sudo apt update
sudo apt install -y git libusb-1.0-0-dev libudev-dev
echo "dtoverlay=dwc2,dr_mode=peripheral" | sudo tee -a /boot/config.txt
echo "dwc2" | sudo tee -a /etc/modules
echo "libcomposite" | sudo tee -a /etc/modules
echo "usb_f_rndis" | sudo tee -a /etc/modules

git config pull.rebase false
git clone https://github.com/Berny23/LD-ToyPad-Emulator.git
cd LD-ToyPad-Emulator

printf '\necho "$UDC" > UDC\nsleep 2;\nchmod a+rw /dev/hidg0' >> usb_setup_script.sh

sudo cp usb_setup_script.sh /usr/local/bin/toypad_usb_setup.sh
sudo chmod +x /usr/local/bin/toypad_usb_setup.sh
(sudo crontab -l 2>/dev/null; echo "@reboot sudo /usr/local/bin/toypad_usb_setup.sh") | sudo crontab -
```
Reboot the raspberry pi with this command:

` sudo shutdown -r now`

Wait a minute or so, until it rebooted. Then connect via SSH again and run the following commands:

```
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.38.0/install.sh | bash
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"

nvm install 11
sudo setcap cap_net_bind_service=+ep `readlink -f \`which node\``
npm install --global node-gyp@8.4.1
npm config set node_gyp $(npm prefix -g)/lib/node_modules/node-gyp/bin/node-gyp.js

cd LD-ToyPad-Emulator
npm install
```

**Usage**

Connect the Raspberry Pi with Your console using a USB cable.

Run the emulator server with this command if you are in the correct folder (otherwise run `cd LD-ToyPad-Emulator` first):

`node index.js`

Type your single board computer's IP address(or "raspberrypi.local") in a browser on your pc or phone to use the emulator.

If you want to turn it off, just press Ctrl + C in the cmd window, then use the command `sudo shutdown now` to safely power off the device.

**Comment by Berny23** [comment]
Thank you!

**Comment by FrankTheTank908** [comment]
Thank you, I’ll test this out with other pi’s I have 😁

**Comment by SubatomicPlanets** [comment]
Thanks for the replies! I also set my Raspberry Pi's hostname to "toypad" and made index.js start after boot. Now every time I plug in my Raspberry Pi (which I never really used for anything else) into my PS4, I can access "toypad.local" on any device connected to the same network, and it works! My real toypad broke a few years ago, and it's really great to play the game again! Thanks to all the people who contributed to this!

**Reply by FrankTheTank908** [reply]
I set mine to toypi 😂

how can I set it to auto start? Mind helping me out? Or us 😂

**Reply by SubatomicPlanets** [reply]
Yeah, I was new to it myself and tried the auto start with something called rc.local, but that didn't work... In the end I did it with "systemd" which is a service manager. Basically you create a new service which runs your script. That service is then called at startup. Here is how you do it:   1. Create a systemd service file: 
```
sudo nano /etc/systemd/system/toypi.service
```
2. Add this to the file:   
```
[Unit]
Description=ToyPi
After=network.target

[Service]
ExecStart=<path_to_node> <path_to_your_script>
WorkingDirectory=<directory_of_your_script>
User=<your_username>
Restart=always

[Install]
WantedBy=multi-user.target
```
replace all the <> fields with the correct information.

3. Start the service:   
```
sudo systemctl enable toypi.service
```

4. Reboot with `sudo reboot`
Make sure the path to node, and all the other information is correct. I am currently not at home, and can only look at the exact file contents tomorrow. Just let me know if it didn't work, and i will look at what I did and post it tomorrow.

**Reply by FrankTheTank908** [reply]
Can you please give me a sample of your service you created because I’m having issues starting mine

With your working directory, listed and everything

I think I’m being confused with the replacements and how it supposed to be

I’m using the default, working directories

edit: this is mine that’s not working

[Unit]
Description=ToyPi
After=network.target

[Service]
ExecStart=node index.js
WorkingDirectory=LD-ToyPad-Emulator
User=frank
Restart=always

[Install]
WantedBy=multi-user.target

**Comment by FrankTheTank908** [comment]
After a long time getting familiar with command line

here’s mine working, change user and version if needed

[Unit]
Description=ToyPi
After=network.target

[Service]
ExecStart=/home/frank/.nvm/versions/node/v11.15.0/bin/node /home/frank/LD-ToyPad-Emulator/index.js
WorkingDirectory=/home/frank/LD-ToyPad-Emulator
User=frank
Restart=always

[Install]
WantedBy=multi-user.target

**Comment by ddwbr** [comment]
I'm using an Orangepi Zero 3 and I wasn't having success, I couldn't get past step 5 of the installation. I tested your code and finally got it, thank you very much!!!

**Comment by jaheemalien** [comment]
Has anyone tried this on Raspberry Pi 3B and got it to work. I tried following the steps but keeping getting an error when trying to run npm install

**Reply by jaheemalien** [reply]
This is the error I get when I try to run npm install:
Traceback (most recent call last):
  File "/home/pi/.nvm/versions/node/v11.15.0/lib/node_modules/node-gyp/gyp/gyp_main.py", line 42, in <module>
    import gyp  # noqa: E402
    ^^^^^^^^^^
  File "/home/pi/.nvm/versions/node/v11.15.0/lib/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 9, in <module>
    import gyp.input
  File "/home/pi/.nvm/versions/node/v11.15.0/lib/node_modules/node-gyp/gyp/pylib/gyp/input.py", line 19, in <module>
    from distutils.version import StrictVersion
ModuleNotFoundError: No module named 'distutils'
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/home/pi/.nvm/versions/node/v11.15.0/lib/node_modules/node-gyp/lib/configure.js:259:16)
gyp ERR! stack     at ChildProcess.emit (events.js:193:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:255:12)
gyp ERR! System Linux 6.12.75+rpt-rpi-v8
gyp ERR! command "/home/pi/.nvm/versions/node/v11.15.0/bin/node" "/home/pi/.nvm/versions/node/v11.15.0/lib/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/pi/LD-ToyPad-Emulator/node_modules/@serialport/bindings
gyp ERR! node -v v11.15.0
gyp ERR! node-gyp -v v8.4.1
gyp ERR! not ok
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: @rollup/rollup-linux-x64-gnu@4.9.5 (node_modules/pad/node_modules/@rollup/rollup-linux-x64-gnu):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for @rollup/rollup-linux-x64-gnu@4.9.5: wanted {"os":"linux","arch":"x64"} (current: {"os":"linux","arch":"arm64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @serialport/bindings@9.2.8 install: `prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @serialport/bindings@9.2.8 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/pi/.npm/_logs/2026-03-29T06_54_41_384Z-debug.log

**Reply by SubatomicPlanets** [reply]
It's been a while since I wrote this post, so I can't help much here. But since it says `ModuleNotFoundError: No module named 'distutils'` in the log, maybe running `pip install setuptools` or using an older Python version might work? If you followed the instructions exactly, then it can pretty much only be the Pi 3B hardware itself, or (more likely) the actual OS you installed on it (the newer ones might come with newer versions of Python and other software). You can maybe try to use a different image in the Raspberry Pi Imager and see if that works (use the newest version of the imager and try older images like "Raspberry Pi OS Legacy 32-bit")

**Reply by jaheemalien** [reply]
So I tried all the commands on the current pi OS and it installed but now when I run node index.js I get this error now:
fs.js:119
    throw err;
    ^

Error: ENOENT: no such file or directory, open '/dev/hidg0'
    at Object.openSync (fs.js:448:3)
    at new RawTransport (/home/pi/LD-ToyPad-Emulator/node_modules/node-ld/dist/transports/raw.js:11:27)
    at new ToyPadEmu (/home/pi/LD-ToyPad-Emulator/node_modules/node-ld/dist/lib/ToyPadEmu.js:65:69)
    at Object.<anonymous> (/home/pi/LD-ToyPad-Emulator/index.js:24:12)
    at Module._compile (internal/modules/cjs/loader.js:816:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:827:10)
    at Module.load (internal/modules/cjs/loader.js:685:32)
    at Function.Module._load (internal/modules/cjs/loader.js:620:12)
    at Function.Module.runMain (internal/modules/cjs/loader.js:877:12)
    at internal/main/run_main_module.js:21:11

**Reply by SubatomicPlanets** [reply]
Not quite sure about this, but `Error: ENOENT: no such file or directory, open '/dev/hidg0'` probably means something is wrong with the USB connection. Are you using a USB cable to plug the Pi into a console like the PS4? I remember that the micro USB port on the Pi didn't work. I had to use a USB-A to USB-A cable. You can also try different USB ports or different cables and see if those work. If none of this works, then it might be the raspberry pi OS version (new ones have the config file at `/boot/firmware/config.txt` while the legacy ones have it at `/boot/config.txt`, so the command from my instructions `echo "dtoverlay=dwc2,dr_mode=peripheral" | sudo tee -a /boot/config.txt` won't work on the new Raspberry Pi OS. You can try running it with the new path, but there might be other issues with the new OS so in that case just use the legacy one).

**Reply by jaheemalien** [reply]
I switched to Legacy boot and Im still getting the same error when running node index.js. I am also trying on PS3 but I think that Raspberry Pi 3B is not compatible so I might just buy a Pi Zero W

**Reply by SubatomicPlanets** [reply]
I just searched if the 3B supports USB gadget mode and you are right, it doesn't. I thought it would work because it's pretty similar to the 3A+ in many ways. Pi Zero W should work though.
